src/lib/cloud/digitalocean/provisioner.ts

```diff
@@ -33,13 +31,89 @@ const REDIRECT_URIS: [string] = [
   //  'http://localhost:10101'
 ];
 
+const STORAGE_KEY_OAUTH = 'DigitalOcean-OAuth';
+
 interface KeyPair {
   private: string;
   public: string;
 }
 
 class Provisioner {
-  constructor(private dispatch_: Function, private state_: any = {}) {}
+  constructor(
+    private sshKeyPair_ :any = {},
+    private network_ :any = {},
+    private storage_ = freedom['core.storage'](),
+
+    // TODO: give this concept a better name, or use a different abstraction?
+    private cloud_ :any = null,
+
+    // doOAuth_() may be called multiple times concurrently before an OAuth
+    // flow has been completed. In order to ensure that each call doesn't
+    // trigger an additional OAuth flow in a new tab, the promise returned by
+    // the first call is cached in this field, and this cached promise is
+    // returned for subsequent calls.
+    private promiseOAuth_ :Promise<Object> = null
+  ) {}
+
+  /*
+   * Returns whether a DO OAuth token is saved in storage.
+   */
+  public hasOAuth = () : Promise<boolean> => {
+    return new Promise((F, R) => {
+      this.getOAuthFromStorage_().then(
```

Comment:
Since this only calls a function which returns a Promise, you don't need to create a "wrapping" Promise:
```typescript
public hasOAuth = () : Promise<boolean> => {
  return this.getOAuthFromStorage_().then((result) => {
    F(!!result);
  });
}
```
No need for an explicit catch handler either, right?